flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -0,0 +1,273 @@
+			// in ExecutionVertexSchedulingRequirements instead of calculating it here
+			CompletableFuture<LogicalSlot> slotFuture =
+					calculatePreferredLocations(execution, verticesScheduledTogether).thenCompose(
+							(Collection<TaskManagerLocation> preferredLocations) ->
+								slotProvider.allocateSlot(
+									slotRequestId,
+									new ScheduledUnit(
+											null,
+											executionVertexId.getJobVertexId(),
+											slotSharingGroupId,
+											schedulingRequirements.getCoLocationConstraint()),
+									new SlotProfile(
+											schedulingRequirements.getResourceProfile(),
+											preferredLocations,
+											Arrays.asList(schedulingRequirements.getPreviousAllocationId()),
+											allPreviousAllocationIds),
+									executionGraph.isQueuedSchedulingAllowed(),
+									executionGraph.getAllocationTimeout()));
+
+			SlotExecutionVertexAssignment slotExecutionVertexAssignment =
+					new SlotExecutionVertexAssignment(executionVertexId, slotFuture);
+			// add to map first to avoid the future completed before added.
+			pendingSlotAssignments.put(executionVertexId, slotExecutionVertexAssignment);
+
+			slotFuture.whenComplete(
+					(ignored, throwable) -> {
+						pendingSlotAssignments.remove(executionVertexId);
+						if (throwable != null) {
+							slotProvider.cancelSlotRequest(slotRequestId, slotSharingGroupId, throwable);
+						}
+					});
+
+			slotExecutionVertexAssignments.add(slotExecutionVertexAssignment);
+		}
+
+		return slotExecutionVertexAssignments;
+	}
+
+	@Override
+	public void cancel(ExecutionVertexID executionVertexId) {
+		SlotExecutionVertexAssignment slotExecutionVertexAssignment = pendingSlotAssignments.get(executionVertexId);
+		if (slotExecutionVertexAssignment != null) {
+			slotExecutionVertexAssignment.getLogicalSlotFuture().cancel(false);
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> stop() {
+		List<ExecutionVertexID> executionVertexIds = new ArrayList<>(pendingSlotAssignments.keySet());
+		executionVertexIds.stream().forEach(executionVertexID -> cancel(executionVertexID));
+
+		return CompletableFuture.completedFuture(null);
+	}
+
+	private static Set<ExecutionVertexID> getAllExecutionVerticesId(
+			Collection<ExecutionVertexSchedulingRequirements> executionVertexSchedulingRequirements) {
+		Set<ExecutionVertexID> executionVertexIds = new HashSet<>(executionVertexSchedulingRequirements.size());
+		for (ExecutionVertexSchedulingRequirements schedulingRequirements : executionVertexSchedulingRequirements) {
+			executionVertexIds.add(schedulingRequirements.getExecutionVertexId());
+		}
+		return executionVertexIds;
+	}
+
+	/**
+	 * Calculates the preferred locations for an execution.
+	 * It will first try to use preferred locations based on state,
+	 * if null, will use the preferred locations based on inputs.
+	 */
+	private static CompletableFuture<Collection<TaskManagerLocation>> calculatePreferredLocations(
+			Execution execution,
```

Comment:
In the original design we are aiming at not having direct dependencies on the `ExecutionGraph` from anywhere but the Scheduler itself:

`getPreferredLocationsBasedOnState()`: If I am not mistaken this can be passed via `ExecutionVertexSchedulingRequirements`

_number of input sources_ & _number of `InputEdges`_: We can pass something similar to a `SchedulingTopology` to the `ExecutionSlotAllocator` so that these information can be inferred. 

What do you think?

(this is not a review)

Reply:
+1 for getting rid of ExecutionGraph in ExecutionSlotAllocator. I had to add it to ExecutionSlotAllocator since we can not getCurrentTaskManagerLocationFuture from SchedulingTopology, could we add this method to SchedulingExecutionVertex?

Reply:
Another problem is that now when calling getConsumedResultPartitions on SchedulingExecutionVertex, we get a Collection<SchedulingResultPartition>, it can't identify how many JobEdges they belong to easily. The existing code will ignore the JobEdge which has too many different location of ExecutionEdge. So could we change Collection<SchedulingResultPartition> getConsumedResultPartitions to Collection<Collection<SchedulingResultPartition>> getConsumedResultPartitions  or that we don't distinguish different JobEdges ?

Reply:
Let's think first which operations we need to implement the status quo:

```
    // gets predecessors (can group by JobVertexID to find predecessors per job vertex)
	Collection<ExecutionVertexID> getConsumedResultPartitionsProducers(ExecutionVertexID executionVertexId);

	// to lookup predecessor TM locations
	Optional<TaskManagerLocation> getTaskManagerLocation(ExecutionVertexID executionVertexId);

	// to filter by >= MAX_DISTINCT_LOCATIONS_TO_CONSIDER
	int getJobVertexParallelism(JobVertexID jobVertexId);
```

Did I miss something?

Reply:
It is decided by the logic we choose for the calculation. Maybe need to get the status of upstream. I create a [doc](https://docs.google.com/document/d/120L_u3C-PGa4jbiEexpFccVToZMowEI7xN-5FMhReZo/edit?usp=sharing) to record some proposal for it.